astropy/units/format/base.py

```diff
@@ -151,34 +149,46 @@ def to_string(
         # First the scale.  Normally unity, in which case we omit
         # it, but non-unity scale can happen, e.g., in decompositions
         # like u.Ry.decompose(), which gives "2.17987e-18 kg m2 / s2".
-        if unit.scale == 1:
-            s = ""
-        else:
-            s = cls.format_exponential_notation(unit.scale)
-
-        # Now the unit baes, taking care that dimensionless does not have any
-        # (but can have a scale; e.g., u.percent.decompose() gives "0.01").
-        if len(unit.bases):
-            if s:
-                s += cls._scale_unit_separator
-            if fraction:
-                numerator, denominator = utils.get_grouped_by_powers(
-                    unit.bases, unit.powers
-                )
-            else:
-                numerator = list(zip(unit.bases, unit.powers))
-                denominator = []
-            if len(denominator):
-                if len(numerator):
-                    numerator = cls._format_unit_list(numerator)
-                else:
-                    numerator = "1"
-                denominator = cls._format_unit_list(denominator)
-                s = cls._format_fraction(s, numerator, denominator, fraction=fraction)
+        s = "" if unit.scale == 1 else cls.format_exponential_notation(unit.scale)
+
+        # dimensionless does not have any bases, but can have a scale;
+        # e.g., u.percent.decompose() gives "0.01".
+        if not unit.bases:
+            return s
+
+        if s:
+            s += cls._scale_unit_separator
+        # Unit powers are monotonically decreasing
+        if not fraction or unit.powers[-1] > 0:
+            return s + cls._format_unit_list(zip(unit.bases, unit.powers))
+
+        positive = []
+        negative = []
+        for base, power in zip(unit.bases, unit.powers, strict=True):
+            if power > 0:
+                positive.append((base, power))
             else:
-                s += cls._format_unit_list(numerator)
-
-        return s
+                negative.append((base, -power))
+        try:
```

Comment:
How about avoiding the whole `dict` and doing:
```
fraction_formatter = getattr(cls, f"_format_{fraction if fraction is not True else 'inline'}_formatter", None)
if fraction_formatter is None:
    supported_formats = [parts[1] for attr in cls.__dict__ if len((parts := attr.split("_")) == 3) and parts[0] = "fraction" and parts[2] == "formatter"] 
    <rest of error dealing>
return fraction_formatter(s, cls._format_unit_list(positive) or "1", cls._format_unit_list(negative))
```

Reply:
This suggestion would make it difficult to find where the `_format_inline_fraction()` and `_format_multiline_fraction()` methods are called. Furthermore, many IDEs and also the GitHub web UI make it very simple to jump from where a function is called to where it is defined, but that wont work with your suggestion because the names of the functions being called are not written down in the source code. So what you are suggesting works fine for Python at runtime, but it will not make it easier to understand or to maintain the code.

Reply:
+1 for static introspectability > line # reduction.

Reply:
OK, that makes sense; the current approach is fine by me too.